bioptim/limits/multinode_constraint.py

```diff
@@ -0,0 +1,358 @@
+        super(MultinodeConstraintList, self)._add(
+            option_type=MultinodeConstraint, multinode_constraint=multinode_constraint, phase=-1, **extra_arguments
+        )
+
+    def print(self):
+        """
+        Print the MultinodeConstraintList to the console
+        """
+        raise NotImplementedError("Printing of MultinodeConstraintList is not ready yet")
+
+    def prepare_multinode_constraints(self, ocp) -> list:
+        """
+        Configure all the phase transitions and put them in a list
+
+        Parameters
+        ----------
+        ocp: OptimalControlProgram
+            A reference to the ocp
+
+        Returns
+        -------
+        The list of all the multi_node constraints prepared
+        """
+        full_phase_multinode_constraint = []
+        existing_phases = []
+        for mnc in self:
+
+            idx_phase = mnc.phase_first_idx
+            if mnc.phase_first_idx >= ocp.n_phases or mnc.phase_second_idx >= ocp.n_phases:
+                raise RuntimeError("Phase index of the multinode_constraint is higher than the number of phases")
+            existing_phases.append(idx_phase)
+
+            if mnc.weight:
+                mnc.base = ObjectiveFunction.MayerFunction
+
+            full_phase_multinode_constraint.append(mnc)
+
+        return full_phase_multinode_constraint
+
+
+class MultinodeConstraintFunctions(PenaltyFunctionAbstract):
+    """
+    Internal implementation of the phase transitions
+    """
+
+    class Functions:
+        """
+        Implementation of all the Multi Node Constraint
+        """
+
+        @staticmethod
+        def continuous(multinode_constraint, all_pn):
+            """
+            The most common continuity function, that is state before equals state after
+
+            Parameters
+            ----------
+            multinode_constraint : MultinodeConstraint
+                A reference to the phase transition
+            all_pn: PenaltyNodeList
+                    The penalty node elements
+
+            Returns
+            -------
+            The difference between the state after and before
+            """
+
+            nlp_pre, nlp_post = all_pn[0].nlp, all_pn[1].nlp
+            states_pre = multinode_constraint.states_mapping.to_second.map(nlp_pre.states.cx_end)
+            states_post = multinode_constraint.states_mapping.to_first.map(nlp_post.states.cx)
+
+            if states_pre.shape != states_post.shape:
+                raise RuntimeError(
+                    f"Continuity can't be established since the number of x to be matched is {states_pre.shape} in the "
+                    f"pre-transition phase and {states_post.shape} post-transition phase. Please use a custom "
+                    f"transition or supply states_mapping"
+                )
+
+            return states_pre - states_post
+
+        @staticmethod
+        def cyclic(multinode_constraint, all_pn) -> MX:
```

Comment:
Cyclic does not make much sense to me as if one wants a cyclic constraint, they should call PHASE_TRANSITION.CYCLIC (which would in turn call the proper mulitnode.equality constraint)

Reply:
done